From 8e55e79fe5d11776b0458f473c96ab7b2038c735 Mon Sep 17 00:00:00 2001
From: Mike Sul <mike.sul@foundries.io>
Date: Tue, 20 Dec 2022 11:45:13 +0100
Subject: [PATCH] Fsync layer once extracted to file system

- Sync an overall file system just after layer unpack while the control
  flow is within the layer "change root".

Signed-off-by: Mike <mike.sul@foundries.io>
Signed-off-by: Jose Quaresma <jose.quaresma@foundries.io>
---
 pkg/chrootarchive/diff_unix.go | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/pkg/chrootarchive/diff_unix.go b/pkg/chrootarchive/diff_unix.go
index 9f7c808ba6..79b5b71c46 100644
--- a/pkg/chrootarchive/diff_unix.go
+++ b/pkg/chrootarchive/diff_unix.go
@@ -5,6 +5,7 @@ package chrootarchive // import "github.com/docker/docker/pkg/chrootarchive"
 import (
 	"io"
 	"path/filepath"
+	"syscall"
 
 	"github.com/containerd/containerd/pkg/userns"
 	"github.com/docker/docker/pkg/archive"
@@ -46,6 +47,7 @@ func applyLayerHandler(dest string, layer io.Reader, options *archive.TarOptions
 		_ = unix.Umask(0)
 
 		size, err := archive.UnpackLayer("/", layer, options)
+		syscall.Sync()
 		done <- result{layerSize: size, err: err}
 	})
 	if err != nil {
-- 
2.34.1

